### PR TITLE
use `Stopwatch.GetElapsedTime` for rate limiter

### PR DESCRIPTION
**Pull request by WeihanLi**
use `Stopwatch.GetElapsedTime` to simplify `IdleDuration` for rate limiter

### CODE REVIEW

File: src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/RateLimiterHelper.cs
Comment (martincostello) [inline src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/RateLimiterHelper.cs:13]: Forgot the `return`?

Comment (WeihanLi) [reply src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/RateLimiterHelper.cs]: Thanks

File: src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs
Comment (tannergooding) [inline src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs:40]: Is it worth ifdefing all these places rather than just always using the internal `RateLimiterHelper.GetElapsedTime`?

Comment (WeihanLi) [reply src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs:40]: Did a small benchmark, benchmark code:

```csharp
[SimpleJob]
[MemoryDiagnoser]
public class StopwatchElapsedTimeTest
{
    private static readonly long StartTime = System.Diagnostics.Stopwatch.GetTimestamp();
    
    [Benchmark(Baseline = true)]
    public TimeSpan StopwatchTest()
    {
        return Stopwatch.GetElapsedTime(StartTime);
    }
    
    [Benchmark]
    public TimeSpan WatchGetElapsedTimeTest()
    {
        return StopwatchHelper.WatchGetElapsedTime(StartTime);
    }
    
    [Benchmark]
    public TimeSpan WatchGetElapsedTime2Test()
    {
        return StopwatchHelper.WatchGetElapsedTime2(StartTime);
    }
    
    [Benchmark]
    public TimeSpan CustomGetElapsedTimeTest()
    {
        return StopwatchHelper.CustomGetElapsedTime(StartTime);
    }
    
    [Benchmark]
    public TimeSpan CustomGetElapsedTime2Test()
    {
        return StopwatchHelper.CustomGetElapsedTime2(StartTime);
    }
}

file static class StopwatchHelper
{
    private static readonly double TickFrequency = (double)TimeSpan.TicksPerSecond / Stopwatch.Frequency;

    public static TimeSpan CustomGetElapsedTime(long startTimestamp) => new((long)((Stopwatch.GetTimestamp() - startTimestamp) * TickFrequency));
    
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static TimeSpan CustomGetElapsedTime2(long startTimestamp) => new((long)((Stopwatch.GetTimestamp() - startTimestamp) * TickFrequency));
    
    public static TimeSpan WatchGetElapsedTime(long startTimestamp) => Stopwatch.GetElapsedTime(startTimestamp);
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static TimeSpan WatchGetElapsedTime2(long startTimestamp) => Stopwatch.GetElapsedTime(startTimestamp);
}
```

testing result:

image

Think maybe we could update the `RateLimiterHelper.GetElapsedTime`

```csharp
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static TimeSpan GetElapsedTime(long startTimestamp)
#if NET
       => Stopwatch.GetElapsedTime(startTimestamp);
#else
       => new((long)((Stopwatch.GetTimestamp() - startTimestamp) * TickFrequency));
#endif
```

Thoughts?

Comment (tannergooding) [reply src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs:40]: > Think maybe we could update the RateLimiterHelper.GetElapsedTime like following to simplify the call in the RateLimiter

That is essentially the idea I was suggesting.

Notably we also do not care about time differences on the order of 1-5ns for most scenarios. We care much more about the code being understandable, maintainable, etc.

Comment (WeihanLi) [reply src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs]: updated

**Review by tannergooding — APPROVED** [review]

**Comment by tannergooding** [comment]
LGTM, but should get a secondary sign-off before merge.

File: src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/RateLimiterHelper.cs
Comment (stephentoub) [inline src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/RateLimiterHelper.cs:13]: Is this necessary?

Comment (WeihanLi) [reply src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/RateLimiterHelper.cs:13]: image
From the benchmark, it looks better with it
for benchmark code: https://github.com/WeihanLi/PerformanceTest/blob/master/PerformanceTest/StopwatchElapsedTimeTest.cs

Comment (tannergooding) [reply src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/RateLimiterHelper.cs:13]: I imagine the difference here is primarily general CPU noise and not due to inlining vs not, it's on the order of 1-5 instruction cycles.

The actual codegen is likely identical and the attribute isn't needed.

Comment (WeihanLi) [reply src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/RateLimiterHelper.cs]: Removed

File: src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs
Comment (stephentoub) [inline src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs:35]: Does every call site follow the same pattern of being:
```C#
nullableTimeSpan is null ? null : GetElapsedTime(nullableTimeSpan.Value)
```
? If so, should the helper do that whole thing to avoid the duplication at each call site?

Comment (WeihanLi) [reply src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs]: Good point, thanks

File: src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/RateLimiterHelper.cs
Comment (WeihanLi) [inline src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/RateLimiterHelper.cs:10]: @stephentoub would you think it's a good idea to public the tick frequency in the `Stopwatch`

https://github.com/dotnet/runtime/blob/b199965ab77731ec2269b80a304f2091caa33ba7/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Stopwatch.cs#L27

Comment (stephentoub) [reply src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/RateLimiterHelper.cs:10]: What would you use it for, other than GetElapsedTime, which is already public?

Comment (stephentoub) [reply src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/RateLimiterHelper.cs:10]: Both of those look they're just GetElapsedTime without wrapping in a TimeSpan. Doesn't seem worthwhile.

**Review by stephentoub — APPROVED** [review]
Thanks!

**Pull request merged**
